### PR TITLE
support function rpad

### PR DESCRIPTION
**Pull request by silentred**
Tried to support the function rpad. It goes alright with `make gotest` and `make server`. The rpad query works with my build on mac.

### CODE REVIEW

**Comment by shenli** [comment]
@silentred Thanks for your PR! Please sign the CLA.

**Comment by shenli** [comment]
@silentred Please add typeinfer info in typeinferer.go.
See https://github.com/ngaut/builddatabase/blob/master/tidb/builtin.md

**Comment by silentred** [comment]
weird, I have signed the CLA. Should I close this one and redo a PR?

**Comment by tiancaiamao** [comment]
no, wait for CI pass, and we'll review it.
it would be merged if two or more reviewer think it's ok.

**Comment by coocood** [comment]
@silentred 
Please add typeinfer on this function, like this PR
https://github.com/pingcap/tidb/pull/2249

File: evaluator/builtin_string.go [hunk 1]
```diff
@@ -603,3 +603,37 @@ func trimRight(str, remstr string) string {
...
+
+	if l < 0 || len(padStr) == 0 {
+		d.SetNull()
+		return d, nil
+	}
+
+	if len(str) >= l {
+		d.SetString(str[:l])
+	} else {
+		dest := str + strings.Repeat(padStr, l-len(str))
```
Comment (tiancaiamao) [inline evaluator/builtin_string.go]: I think it's not `strings.Repeat(padStr, l-len(str))` , padStr may not just one char, the total length after `Rpad` is l:

```
mysql> SELECT RPAD('h',5,'ab');
+------------------+
| RPAD('h',5,'ab') |
+------------------+
| habab            |
+------------------+
1 row in set (0.00 sec)
```

**Comment by silentred** [comment]
@shenli @tiancaiamao @coocood Thanks for the advises. I will fix these.

File: evaluator/builtin_string_test.go [hunk 2]
```diff
@@ -707,15 +707,18 @@ func (s *testEvaluatorSuite) TestRpad(c *C) {
 		{"hi", 5, "?", "hi???"},
 		{"hi", 1, "?", "h"},
 		{"hi", -1, "?", ""},
-		{"hi", 1, "", ""},
+		{"hi", 1, "", "h"},
+		{"hi", 5, "", ""},
+		{"hi", 5, "ab", "hiaba"},
+		{"hi", 6, "ab", "hiabab"},
```
Comment (coocood) [inline evaluator/builtin_string_test.go]: The test can't distinguish null and empty string.
`rpad("hi", 0, "?")` is empty string, not NULL.

File: evaluator/builtin_string.go [hunk 3]
```diff
@@ -603,3 +603,49 @@ func trimRight(str, remstr string) string {
...
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+	l := int(length)
+
+	padStr, err := args[2].ToString()
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
```
Comment (tiancaiamao) [inline evaluator/builtin_string.go]: IMHO, the following logic too much complex than it should be, and the loop is  hard to read.
Here is a simpler example:
```
func test(str string, length int, pad string) {
	for len(str) < length {
		str = str + pad
	}
	return str[:length]
}
```

just check `pad != ""` before this logic.

File: evaluator/builtin_string.go [hunk 4]
```diff
@@ -603,3 +603,49 @@ func trimRight(str, remstr string) string {
...
+		return d, nil
+	}
+
+	if l < 0 || len(padStr) == 0 {
+		d.SetNull()
+		return d, nil
+	}
+
+	tailLen := l - len(str)
+	padStrBytes := []byte(padStr)
```
Comment (coocood) [inline evaluator/builtin_string.go]: How about 
```
repeatCount := tailLen/len(padStr) + 1
dest := str + strings.Repeat(padStr, repeatCount)
dest = dest[:l]
```

File: evaluator/builtin_string.go (hunk 3 above)
Comment (silentred) [inline evaluator/builtin_string.go]: Thanks, it is simpler indeed. I worried that `str = str + pad` would allocate memory every time, so I chose pre-allocated []byte.

File: evaluator/builtin_string.go (hunk 4 above)
Comment (silentred) [inline evaluator/builtin_string.go]: Looks better. I will change.

**Comment by coocood** [comment]
@silentred 
Please add a test case in `typeinferer_test.go`

**Comment by coocood** [comment]
LGTM

**Comment by tiancaiamao** [comment]
LGTM

**Comment by coocood** [comment]
@silentred 
Thank you for your contribution!

**Pull request merged**